lib/src/compiler.ts

```diff
@@ -0,0 +1,198 @@
+// Copyright 2023 Google LLC. Use of this source code is governed by an
+// MIT-style license that can be found in the LICENSE file or at
+// https://opensource.org/licenses/MIT.
+
+import * as p from 'path';
+import * as supportsColor from 'supports-color';
+import {deprotofySourceSpan} from './deprotofy-span';
+import {Dispatcher, DispatcherHandlers} from './dispatcher';
+import {Exception} from './exception';
+import {ImporterRegistry} from './importer-registry';
+import {
+  legacyImporterProtocol,
+  removeLegacyImporter,
+  removeLegacyImporterFromSpan,
+} from './legacy/utils';
+import {MessageTransformer} from './message-transformer';
+import * as utils from './utils';
+import * as proto from './vendor/embedded_sass_pb';
+import {SourceSpan} from './vendor/sass';
+import {CompileResult} from './vendor/sass/compile';
+import {Options, StringOptions} from './vendor/sass/options';
+
+// A module-level ID for each compilation.
+let compilationId = 0;
+
+/// Allow the legacy API to pass in an option signaling to the modern API that
+/// it's being run in legacy mode.
+///
+/// This is not intended for API users to pass in, and may be broken without
+/// warning in the future.
+export type OptionsWithLegacy<sync extends 'sync' | 'async'> = Options<sync> & {
+  legacy?: boolean;
+};
+
+/// Allow the legacy API to pass in an option signaling to the modern API that
+/// it's being run in legacy mode.
+///
+/// This is not intended for API users to pass in, and may be broken without
+/// warning in the future.
+export type StringOptionsWithLegacy<sync extends 'sync' | 'async'> =
+  StringOptions<sync> & {legacy?: boolean};
+
+/**
+ * Creates a dispatcher that dispatches messages from the given `stdout` stream.
+ */
+export function createDispatcher<sync extends 'sync' | 'async'>(
+  messageTransformer: MessageTransformer,
+  handlers: DispatcherHandlers<sync>
+): Dispatcher<sync> {
+  return new Dispatcher<sync>(
+    (compilationId += 1),
```

Comment:
This is probably ok for most of use cases, but can be broken if sass is kept running. E.g. If I have a live-reload web server that keep compiling changes, eventually this number will become 0xFFFFFFFF or higher, which is not allowed by protocol.

In ruby host this issue is handled in two ways:

1. The `compilationId` is reset to `1` whenever the compiler becomes idle (no more pending compilations). - Also see the other comment above, `compilationId` should be tracked per compiler, not globally.
2. If id keep grow and actually hit `0xFFFFFFFF`, replace the compiler process with a new one, and async dispose the current compiler process. Actual implementation of async dispose is quite different, but logically it's same that it is to wait for all pending compilations to finish and clean up all the resources.

The case that the 2nd item handles would be very unlikely to happen if 1st item is implemented. The 2nd item itself is quite difficult to implement, therefore I think we should at least implement the 1st item.

Reply:
I agree that we should implement the first, I think we can get away without implementing the second.

Reply:
Implemented per-compiler IDs in 297cf6734b082823eeedf493f39971eb24d97ad6. It seemed like an important enough detail, so I added unit tests as well